File: spring_mvc_jpa_parallel_virtual_thread/src/main/java/me/yeonhyuk/heavy_traffic/spring_mvc_jpa_parallel_virtual_thread/common/GlobalResponse.java
package me.yeonhyuk.heavy_traffic.spring_mvc_jpa_parallel_virtual_thread.common;

/**
 * 모든 API 가 동일한 형태로 반환하는 응답 DTO (SR §5).
 * resultData 는 제네릭 타입으로 정의한다.
 */
public record GlobalResponse<T>(ResultCodeEnum resultCode, T resultData) {

    public static <T> GlobalResponse<T> success(T resultData) {
        return new GlobalResponse<>(ResultCodeEnum.SUCCESS, resultData);
    }

    public static <T> GlobalResponse<T> error(ResultCodeEnum resultCode) {
        return new GlobalResponse<>(resultCode, null);
    }
}

File: nestjs_drizzle/src/common/global-response.ts
import { ResultCode } from './result-code.enum';

/**
 * 모든 API 가 동일한 형태로 반환하는 표준 응답 DTO (SR §5).
 * resultData 는 제네릭 타입이며, 오류 시 null 이다.
 */
export class GlobalResponse<T> {
  constructor(
    readonly resultCode: number,
    readonly resultData: T | null,
  ) {}

  static success<T>(resultData: T): GlobalResponse<T> {
    return new GlobalResponse<T>(ResultCode.SUCCESS, resultData);
  }

  static error(resultCode: number): GlobalResponse<null> {
    return new GlobalResponse<null>(resultCode, null);
  }
}
